### PR TITLE
*: support resource_group sql hint

### LINKED ISSUE #41676: support resource_group hint for resource control feature
## Enhancement

To support the flexibility of set different resource group from the one bound to the user, TiDB needs to support resource_group hint as MySQL does.  It's an enhancement of issue #38825

### PR DESCRIPTION
**Pull request by BornChanger**
### What problem does this PR solve?

Issue Number: close #41676 

Problem Summary:

### What is changed and how it works?

Via this pr, resource control can support resource _group hint, here is an example

select /*+ resource_group(rg1) */ * from t1;

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)
```sql
mysql> select * from information_schema.resource_groups;
+---------+------------+-----------+
| NAME    | RU_PER_SEC | BURSTABLE |
+---------+------------+-----------+
| default |   10000000 | YES       |
| rg1     |      10000 | NO        |
+---------+------------+-----------+
2 rows in set (0.00 sec)

mysql> select /*+ resource_group(rg1) */ * from t1;
Empty set (0.01 sec)

mysql> select /*+ resource_group(rg2) */ * from t1;
ERROR 1105 (HY000): [resource group] resourceGroupName rg2 is not existed.

mysql> insert /*+ resource_group(rg2) */ into t1 values(1);
ERROR 1105 (HY000): [resource group] resourceGroupName rg2 is not existed.

mysql> delete /*+ resource_group(rg3) */ from t1;
ERROR 1105 (HY000): [resource group] resourceGroupName rg3 is not existed.
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/resource_group_test.go [hunk 1]
```diff
@@ -176,3 +176,21 @@ func testResourceGroupNameFromIS(t *testing.T, ctx sessionctx.Context, name stri
 	g, _ := dom.InfoSchema().ResourceGroupByName(model.NewCIStr(name))
 	return g
 }
+
+func TestResourceGroupHint(t *testing.T) {
+	store, _ := testkit.CreateMockStoreAndDomain(t)
```
Comment (hawkingrei) [inline ddl/resource_group_test.go:181]:
```suggestion
	store := testkit.CreateMockStore(t)
```

File: ddl/resource_group_test.go [hunk 2]
```diff
@@ -176,3 +176,21 @@ func testResourceGroupNameFromIS(t *testing.T, ctx sessionctx.Context, name stri
 	g, _ := dom.InfoSchema().ResourceGroupByName(model.NewCIStr(name))
```
Comment (hawkingrei) [inline ddl/resource_group_test.go:176]: Please move it into ```ddl/resourcegrouptest```If possible.

the test cases in the ddl are too large to run as soon as possible.

Comment (hawkingrei) [reply ddl/resource_group_test.go:176]: ref #41377

File: planner/optimize.go [hunk 3]
```diff
@@ -162,6 +162,12 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 		}
 	}
 
+	// Override the resource group if necessary
+	// TODO: we didn't check the existence of the hinted resource group now to save the cost per query
+	if originStmtHints.HasResourceGroup {
+		sessVars.ResourceGroupName = originStmtHints.ResourceGroup
```
Comment (glorv) [inline planner/optimize.go:168]: What about use put the resource group hint into StatementContext, so we don't need to change and restore the session variable.

File: ddl/resourcegrouptest/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,18 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_test")
+
+go_test(
+    name = "resourcegrouptest_test",
+    srcs = ["resource_group_test.go"],
```
Comment (hawkingrei) [inline ddl/resourcegrouptest/BUILD.bazel:5]: add

```
    timeout = "short",
    flaky = True,
```

File: planner/optimize.go (hunk 3 above)
Comment (BornChanger) [reply planner/optimize.go:168]: As we talked offline, your proposed change doesn't simplify the work.

File: session/session.go [hunk 5]
```diff
@@ -2236,6 +2239,9 @@ func (s *session) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlex
 		metrics.ResourceGroupQueryTotalCounter.WithLabelValues(resourceGroupName).Inc()
 	}
 
+	// Restore the resource group for the session
+	s.GetSessionVars().ResourceGroupName = originalResourceGroup
```
Comment (glorv) [inline session/session.go:2243]: Better to make this a defer before `compiler.Compile` to ensure the session var can be reset back even if this function return early when meets error.

Comment (BornChanger) [reply session/session.go]: Good suggestion.

**Review by glorv — APPROVED** [review]
LGTM

File: planner/optimize.go [hunk 6]
```diff
@@ -162,6 +162,12 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 		}
 	}
 
+	// Override the resource group if necessary
+	// TODO: we didn't check the existence of the hinted resource group now to save the cost per query
+	if originStmtHints.HasResourceGroup {
```
Comment (nolouch) [inline planner/optimize.go:167]: Do we need to check if the resource control is enabled?

Comment (BornChanger) [reply planner/optimize.go]: check added

File: planner/optimize.go [hunk 7]
```diff
@@ -162,6 +162,15 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 		}
 	}
 
+	// Override the resource group if necessary
+	// TODO: we didn't check the existence of the hinted resource group now to save the cost per query
+	if originStmtHints.HasResourceGroup && variable.EnableResourceControl.Load() {
```
Comment (nolouch) [inline planner/optimize.go:167]: I think more readable and can reduce once if judged for normal situations.
```
if originStmtHints.HasResourceGroup {
       if variable.EnableResourceControl.Load() {
       }
       ...
}
```

Comment (BornChanger) [reply planner/optimize.go]: fixed

**Review by tiancaiamao — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Pull request merged**